odl/contrib/param_opt/param_opt.py

```diff
@@ -0,0 +1,92 @@
+# Copyright 2014-2017 The ODL contributors
+#
+# This file is part of ODL.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Method(s) to find optimal reconstruction parameter(s) w.r.t. given FOM."""
+
+import numpy as np
+import scipy.optimize
+
+__all__ = ('optimal_parameters', )
+
+
+def optimal_parameters(reconstruction, fom, phantoms, data,
+                       initial_param=0):
+    r"""Find the optimal parameters for a reconstruction method.
+
+    Notes
+    -----
+    For a forward operator :math:`A : X \to Y`, a reconstruction operator
+    parametrized by :math:`\theta` is some operator
+    :math:`R_\theta : Y \to X`
+    such that
+
+    .. math::
+        R_\theta(A(x)) \approx x.
+
+    The optimal choice of :math:`\theta` is given by
+
+    .. math::
+        \theta = \arg\min_\theta fom(R(A(x) + noise), x)
+
+    where :math:`fom : X \times X \to \mathbb{R}` is a figure of merit.
+
+    Parameters
+    ----------
+    reconstruction : callable
+        Function that takes two parameters:
+
+            * data : The data to be reconstructed
+            * parameters : Parameters of the reconstruction method
+
+        The function should return the reconstructed image.
+    fom : callable
+        Function that takes two parameters:
+
+            * reconstructed_image
+            * true_image
+
+        and returns a scalar figure of merit.
+    phantoms : sequence
+        True images.
+    data : sequence
+        The data to reconstruct from.
+    initial_param : array-like, optional
+        Initial guess for the parameters, default is zero.
+
+    Returns
+    -------
+    parameters : 'numpy.ndarray'
+        The  optimal parameters for the reconstruction problem.
+    """
+
+    def func(lam):
+        # Function to be minimized by scipy
+        return sum(fom(reconstruction(datai, lam), phantomi)
+                   for phantomi, datai in zip(phantoms, data))
+
+    # Pick resolution to fit the one used by the space
+    tol = np.finfo(phantoms[0].space.dtype).resolution * 10
+
+    initial_param = np.asarray(initial_param)
+
+    # We use a faster optimizer for the one parameter case
+    if initial_param.size == 1:
+        bracket = [initial_param - tol, initial_param + tol]
```

Comment:
Since `tol` is typically ridiculously small (1e-14), won't this bracket also be ridiculously narrow?

Reply:
Sure will, but that's fine if we read the doc:

```
bracket : sequence, optional
    For methods 'brent' and 'golden', `bracket` defines the bracketing
    interval and can either have three items ``(a, b, c)`` so that
    ``a < b < c`` and ``fun(b) < fun(a), fun(c)`` or two items ``a`` and
    ``c`` which are assumed to be a starting interval for a downhill
    bracket search (see `bracket`); it doesn't always mean that the
    obtained solution will satisfy ``a <= x <= c``.
```

With that said, users should probably not need to read the doc, and this should me mentioned in a comment.

Reply:
There's also a single backtick issue with \`bracket\` while we're at it. And `'brent'` and `'golden'` should be double-backticked as well.

Reply:
This is litterally copy paste from scipy :D

Reply:
You are right, the bracket could be anything, and that bracket, or the default one, works for some nice functions. However, I did some tests on Shepp-Logan+FPB(cutoff)+SSIM, and, in that case, that tiny bracket really seems to be a problem: the minimiser always returns the initial value. Note that the default bracket does not work either (but the result is different). The function to minimise looks like ~the Texas border~ [that](https://imgur.com/nRUTDtW).

Reply:
Perhaps we should allow `initial_param` to be either a point value or a bracket then!

Reply:
Ignoring this for now, made an issue: 1422